### PR TITLE
Jakobj/fix/setup compilechannels

### PR DESCRIPTION
**Pull request by WillemWybo**
@jakobj somehow can't see your repository so have to make a pull request to the main repository.

I solved the install problem by adding the install commands explicitly again.

For the neuron initialization, I kept running into the problem that if you run it after calling `python3 setup.py install` from the same directory, `compilechannels` imports the `neat` from that directory, hence the compiled files are not in the install location. Solved it in a not-so-nice way by creating a new script `initneuron` in `neat/tools/simtools/neuron/` that you call after install. Not sure if there is a better solution...

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -18,33 +19,28 @@
 from __version__ import version as pversion
 
 
-"""
-Define pre- and post-install commands via command classes.
-From https://stackoverflow.com/questions/20288711/post-install-script-with-python-setuptools
-"""
-
-class PostDevelopCommand(develop):
+class DevelopCommand(develop):
```
Comment (jakobj) [inline setup.py:22]: this could be removed, right?

File: setup.py [hunk 2]
```diff
@@ -18,33 +19,28 @@
...
-        write_ionchannel_header_and_cpp_file()
-        copy_default_neuron_mechanisms()
         # develop install
         develop.run(self)
-        # execute post installation commands
-        compile_default_ion_channels()
 
 
-class PostInstallCommand(install):
+class InstallCommand(install):
```
Comment (jakobj) [inline setup.py:29]: (see above)

File: setup.py [hunk 3]
```diff
@@ -109,24 +75,26 @@ def read_requirements():
...
               'neat.channels.channelcollection'],
+    package_data={
+        "neat.tools.simtools.neuron": ["mech/*.mod"],
+    },
     ext_package='neat',
     ext_modules=[ext],
     cmdclass={
-        'develop': PostDevelopCommand,
-        'install': PostInstallCommand,
+        'develop': DevelopCommand,
```
Comment (jakobj) [inline setup.py:94]: (see above)

File: neat/channels/compilechannels [hunk 4]
```diff
@@ -21,10 +28,11 @@ def allBaseClasses(cls):
 if not os.path.exists(path_for_mod_files):
         os.makedirs(path_for_mod_files)
 
-for channel_module in glob.glob(path_to_dir + '*.py'):
+sys.path.insert(0, path_for_channels)
+for channel_module in glob.glob(path_for_channels + '*.py'):
```
Comment (jakobj) [inline neat/channels/compilechannels:32]:
```suggestion
for channel_module in glob.glob(os.path.join(path_for_channels, '*.py')):
```

File: setup.py (hunk 3 above)
Comment (WillemWybo) [reply setup.py:94]: This is the only way I got the toolbox properly in my install folder. Don't know what the problem is but at least if this one works for pip and setup.py and linux and mac we should keep it for now, even though it is ugly.

**Comment by WillemWybo** [comment]
We may also consider that at some point we may also want different operation for neuron initialization than what `compilechannels` does. Hence `initneuron` could be a handy option. It could also reset the neuron install to it's default state, e.g. if users have added ion channels with bugs and want to get rid of them.

**Comment by WillemWybo** [comment]
@jakobj alternative solution for ion channel compilation problem upon install, now it can be called from package root directory. I'm worried that it might break in this pip developer mode?

File: setup.py (hunk 3 above)
Comment (jakobj) [reply setup.py:94]: hmm, this is really strange. the "new" classes just do what would anyway be done. ok, let's leave it for now, maybe create an issue to remind us?

File: setup.py [hunk 5]
```diff
@@ -109,24 +75,26 @@ def read_requirements():
 s_ = setup(
     name='neat',
     version=pversion,
-    scripts=['neat/channels/compilechannels'],
+    scripts=['neat/channels/compilechannels',
+             'neat/tools/simtools/neuron/initneuron'],
```
Comment (jakobj) [inline setup.py:79]: should be removed if we use your new option

File: neat/channels/compilechannels [hunk 6]
```diff
@@ -1,14 +1,27 @@
 #!/usr/bin/env python3
-import os, glob, sys, inspect, shutil, subprocess
+import imp, os, glob, sys, inspect, shutil, subprocess
+
+# get new paths list without the current working directory
```
Comment (jakobj) [inline neat/channels/compilechannels:4]: interesting approach, mabye put this into a function, e.g., `import_neat_from_nonlocal_path`?

File: neat/channels/compilechannels [hunk 7]
```diff
@@ -1,14 +1,27 @@
 #!/usr/bin/env python3
-import os, glob, sys, inspect, shutil, subprocess
+import imp, os, glob, sys, inspect, shutil, subprocess
```
Comment (jakobj) [inline neat/channels/compilechannels:2]: i'm getting a deprecation warning:
```bash
DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
```

**Comment by jakobj** [comment]
> @jakobj alternative solution for ion channel compilation problem upon install, now it can be called from package root directory. I'm worried that it might break in this pip developer mode?

yes, you were right, it breaks in pip develop mode :(

File: neat/channels/compilechannels [hunk 8]
```diff
@@ -1,14 +1,27 @@
...
+
+# get new paths list without the current working directory
+current_path = os.path.realpath(os.getcwd())
+paths = [path for path in sys.path if os.path.realpath(path) != current_path]
```
Comment (jakobj) [inline neat/channels/compilechannels:6]:
```suggestion
paths = [path for path in sys.path if os.path.realpath(path) != current_path]
paths += [current_path]
```

**Comment by jakobj** [comment]
i've added a suggestion that fixes this in develop mode: we can append the current path to the end of the path list. when we are working with a standard installation it will import from the environment path, when in develop mode from the local path because python doesn't find the module anywhere else. i think this is a practical solution for now. @WillemWybo

File: neat/channels/compilechannels (hunk 7 above)
Comment (WillemWybo) [reply neat/channels/compilechannels]: I've committed an implementation with the new importlib

File: neat/channels/compilechannels (hunk 6 above)
Comment (WillemWybo) [reply neat/channels/compilechannels]: put the import in a function

File: setup.py (hunk 5 above)
Comment (WillemWybo) [reply setup.py]: removed initneuron

File: setup.py (hunk 1 above)
Comment (WillemWybo) [reply setup.py:22]: see new issue #75

**Comment by WillemWybo** [comment]
I've used `importlib` now. If you could test it with pip in develop and install modes we can then merge the pull requrest

File: neat/channels/compilechannels [hunk 9]
```diff
@@ -1,14 +1,52 @@
...
+
+    # load module with importlib according to
+    # https://stackoverflow.com/questions/35288021/what-is-the-equivalent-of-imp-find-module-in-importlib
+    # and to
+    # https://import-sig.python.narkive.com/JW8eIKSg/how-best-to-replace-imp-load-module
+    import importlib
+    neatspec = importlib.machinery.PathFinder().find_spec("neat", paths)
+    neat = neatspec.loader.load_module()
+    neatspec.loader.exec_module(neat)
+    # Deprecated implementation with imp package
```
Comment (jakobj) [inline neat/channels/compilechannels:27]: i would remove this, no need to keep outdated code around, git does that for us ;)

Comment (WillemWybo) [reply neat/channels/compilechannels]: removed

**Pull request merged**